### PR TITLE
Fix wrong condition in base::internalize_symbols().

### PR DESCRIPTION
**Pull request by michaelwoerister**
Fix a typo that snuck into https://github.com/rust-lang/rust/pull/34899 (and completely broke `internalize_symbols()`).

### CODE REVIEW

**Comment by eddyb** [comment]
Would be nice if you could add a test, but otherwise r=me.

**Comment by TimNN** [comment]
This fixes #34891 for me.

**Comment by michaelwoerister** [comment]
@TimNN Excellent `:)`

**Comment by TimNN** [comment]
The travis failure looks legit to me.

**Comment by michaelwoerister** [comment]
Yes, the error looks legit. The internalizer does not take the `#[no_mangle]` and explicit linkage attributes into account, something that just never was tested before because it only ran with multiple codegen units enabled. If have a fix for this but I want to test it some more and maybe write a regression test too.

**Comment by michaelwoerister** [comment]
Alright, that seems to have worked. @eddyb, care to review that last commit too?

File: src/librustc_trans/base.rs [hunk 1]
```diff
@@ -2256,12 +2257,20 @@ fn write_metadata(cx: &SharedCrateContext,
...
+
+    // 'unsafe' because we are holding on to CStr's from the LLVM module within
+    // this block.
     unsafe {
-        let mut declared = HashSet::new();
+        let mut referenced_somewhere = FnvHashSet();
 
-        // Collect all external declarations in all compilation units.
-        for ccx in cx.iter() {
+        // Collect all symbols that need stay externally visible because they
```
Comment (michaelwoerister) [inline src/librustc_trans/base.rs]: whoops, typo: "need to stay"

**Comment by eddyb** [comment]
r=me with the typo fixed.

**Comment by alexcrichton** [comment]
@bors: retry

sorry for the number of retries...

**Comment by nikomatsakis** [comment]
@bors p=1

this may fix a huge regr in compiler perf

**Pull request merged**